lake/commits/store.go

```diff
@@ -0,0 +1,279 @@
+package commits
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/brimdata/zed/pkg/storage"
+	"github.com/brimdata/zed/zio"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zngbytes"
+	"github.com/brimdata/zed/zson"
+	"github.com/segmentio/ksuid"
+)
+
+var (
+	ErrBadCommitObject = errors.New("first record of object not a commit")
+	ErrExists          = errors.New("commit object already exists")
+	ErrNotFound        = errors.New("commit object not found")
+)
+
+type Store struct {
+	path   *storage.URI
+	engine storage.Engine
+	// We use these poor-man's caches for now since chasing linked lists
+	// over cloud storage will run very slowly... there is an interesting
+	// scaling problem to work on here.
+	cache     map[ksuid.KSUID]*Object
+	snapshots map[ksuid.KSUID]*Snapshot
+	paths     map[ksuid.KSUID][]ksuid.KSUID
```

Comment:
Use github.com/hashicorp/golang-lru for these.

Reply:
I made issue #3002 for this.  Sizing the caches and configuring limits will need some thought.  And the approach may soon change anyway with a design for persisting snapshots.  This is definitely a scaling issue right now, but totally solvable.